### PR TITLE
Refactor shards into a Feature

### PR DESCRIPTION
**Pull request by fujnw**
Shard wallet and upgrades are both housed inside `Shards` feature for now. The fact that shards need a separate set of currencies makes it hard to just extend the existing Wallet and Upgrade features. I didn't want the added complexity of creating 18 new currencies and 72 upgrade objects just for this.

### CODE REVIEW

File: src/scripts/shards/Shards.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+    public shardWallet: ArrayOfObservables<number>;
+    public shardUpgrades: ArrayOfObservables<number>;
+
+    constructor() {
+        this.shardWallet = new ArrayOfObservables(this.defaults.shardWallet);
+        this.shardUpgrades = new ArrayOfObservables(this.defaults.shardUpgrades);
+    }
+
+    public gainShards(pokemon: BattlePokemon) {
```
Comment (Ishadijcks) [inline src/scripts/shards/Shards.ts]: Shards should not have to know of the pokemons, pass the shard type here

File: src/scripts/shards/Shards.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
+class Shards implements Feature {
+    name: string = "Shards";
+    saveKey: string = "shards";
+
+    defaults = {
+        'shardWallet': Array.apply(null, Array<number>(18)).map(() => 0),
+        'shardUpgrades': Array.apply(null, Array<number>(18 * 4)).map(() => 0),
```
Comment (Ishadijcks) [inline src/scripts/shards/Shards.ts]: Why not an 18 * 4 array?

File: src/scripts/shards/Shards.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+    public getShardUpgradeCost(
+        typeNum: GameConstants.PokemonType, 
+        effectNum: GameConstants.TypeEffectiveness
+    ): number {
+        let cost = (this.getShardUpgrade(typeNum, effectNum) + 1) * GameConstants.SHARD_UPGRADE_COST;
+        return cost;
+    }
+
+    public shardUpgradeMaxed(
```
Comment (Ishadijcks) [inline src/scripts/shards/Shards.ts]: Please prefix boolean functions by `is`, `has`, `can`, etc...

File: src/scripts/shards/Shards.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        effectNum: GameConstants.TypeEffectiveness
+    ) {
+        return this.shardUpgrades[typeNum * 4 + effectNum];
+    }
+
+    initialize() {
+    }
+
+    canAccess(): boolean {
+        return true;
```
Comment (Ishadijcks) [inline src/scripts/shards/Shards.ts]: Should be changed to `App.game.keyItems(KeyItems.KeyItem.Shard_case)`

File: src/scripts/shards/Shards.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            'shardUpgrades': this.shardUpgrades.map(x => x),
+        }
+    }
+
+    fromJSON(json: object) {
+        if (json == null) {
+            return
+        }
+
+        if (json != null) {
```
Comment (Ishadijcks) [inline src/scripts/shards/Shards.ts]: Redundant check

File: src/scripts/shards/Shards.ts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        return true;
+    }
+
+    update(delta: number) {
+    }
+
+    toJSON(): object {
+        return {
+            'shardWallet': this.shardWallet.map(x => x),
+            'shardUpgrades': this.shardUpgrades.map(x => x),
```
Comment (Ishadijcks) [inline src/scripts/shards/Shards.ts]: Clever!

File: src/components/shards.html [hunk 7]
```diff
@@ -8,22 +8,22 @@ <h4 class="modal-title"></h4>
                      src="assets/images/questionmark.png">
 
             </div>
-            <div class="modal-body" id="shardBody" data-bind='foreach: player._shardUpgrades'>
+            <div class="modal-body" id="shardBody" data-bind='foreach: Array(18)'>
```
Comment (Ishadijcks) [inline src/components/shards.html]: Should use the length of the shardEnum, maybe from a `ShardController.getDisplayableShards()`

File: src/components/shards.html [hunk 8]
```diff
@@ -8,22 +8,22 @@ <h4 class="modal-title"></h4>
...
                 <h3>
                     <span data-bind="text: GameConstants.PokemonType[$index()]"></span>
                     <div style="float:right;">
                       <img data-bind='attr: { src: "assets/images/shards/" + $index() + ".png" }'/>
-                      <i data-bind="text: player._shardsCollected[$index()]().toLocaleString('en-US')"></i>
+                      <i data-bind="text: App.game.shards.shardWallet[$index()].toLocaleString('en-US')"></i>
                     </div>
                 </h3>
-                <table class='table' data-bind='foreach: $data'>
+                <table class='table' data-bind='foreach: Array(4)'>
```
Comment (Ishadijcks) [inline src/components/shards.html]: Should use some variable here

File: src/scripts/shards/Shards.ts (hunk 2 above)
Comment (fujnw) [reply src/scripts/shards/Shards.ts]: Our `ArrayOfObservables` is designed to work on 1d arrays. I didn't want to spend the time to make a 2D array observe correctly.

File: src/scripts/shards/Shards.ts (hunk 6 above)
Comment (fujnw) [reply src/scripts/shards/Shards.ts]: Probably be good to have this "convert to regular arrays" function in the `ArrayOfObservables` itself.

**Comment by fujnw** [comment]
With the way I implemented the shard button on the game menu, the Shard button hover shows the cursor, while on all other buttons, the cursor becomes a hand. Any way to make the hover on Shards also a hand?

**Comment by Aegyo** [comment]
You can add a `href="#"` attribute to your menu link, or add some css somewhere to apply `cursor: pointer` to all of the `li` elements in the menu

File: src/components/gameMenu.html [hunk 9]
```diff
@@ -17,7 +17,7 @@
             <a class="dropdown-item" href="#achievementsModal" data-toggle="modal">Achievements</a>
         </li>
         <li>
-            <a class="dropdown-item" href="#shardModal" data-toggle="modal">Shards</a>
+            <a class="dropdown-item" data-bind="click:function(){App.game.shards.openShardModal()}" data-toggle="modal">Shards</a>
```
Comment (RedSparr0w) [inline src/components/gameMenu.html]:
```suggestion
            <a class="dropdown-item" href="#" data-bind="click:function(){App.game.shards.openShardModal()}">Shards</a>
```

**Pull request closed without merging**